python/sglang/srt/mem_cache/memory_pool.py

```diff
@@ -162,23 +224,81 @@ def get_value_buffer(self, layer_id: int):
     def get_kv_buffer(self, layer_id: int):
         return self.get_key_buffer(layer_id), self.get_value_buffer(layer_id)
 
+    def get_key_scales_buffer(self, layer_id: int):
+        if self.dtype == torch.int8:
+            return self.k_scales_buffer[layer_id]
+        else:
+            return None
+
+    def get_value_scales_buffer(self, layer_id: int):
+        if self.dtype == torch.int8:
+            return self.v_scales_buffer[layer_id]
+        else:
+            return None
+
+    def get_kv_scales_buffer(self, layer_id: int):
+        if self.dtype == torch.int8:
+            return self.get_key_scales_buffer(layer_id), self.get_value_scales_buffer(
+                layer_id
+            )
+        else:
+            return None
+
     def set_kv_buffer(
         self,
         layer_id: int,
         loc: torch.Tensor,
         cache_k: torch.Tensor,
         cache_v: torch.Tensor,
     ):
-        if cache_k.dtype != self.dtype:
-            cache_k = cache_k.to(self.dtype)
-        if cache_v.dtype != self.dtype:
-            cache_v = cache_v.to(self.dtype)
-        if self.store_dtype != self.dtype:
-            self.k_buffer[layer_id][loc] = cache_k.view(self.store_dtype)
-            self.v_buffer[layer_id][loc] = cache_v.view(self.store_dtype)
+        if self.dtype == torch.int8:
+            if self.kv_cache_dtype_str == "int4":
+                from sglang.srt.layers.attention.triton_ops.decode_attention_int4kv import (
+                    destindex_copy_quantize_int4kv,
```

Comment:
do we want to pre-cache this function to avoid the overhead of if/else and import?

Reply:
fix import